### ISSUE TITLE
Continuous diff in aws_s3_bucket_object content

### ISSUE THREAD

**Issue by ThePletch** [hashicorp/terraform#15594]
When creating an aws_s3_bucket_object from the raw content of a zip file, Terraform reports a diff for every plan due to what appears to be an encoding issue.

Use case explanation: passing around the content of an archive file rather than its file path prevents issues caused by storing absolute file paths in Terraform state (see https://github.com/hashicorp/terraform/issues/7927).

### Terraform Version
Terraform v0.9.6 (fix for this issue does not appear in newer changelogs)

### Terraform Configuration Files
```hcl
data "archive_file" "zip" {
  type = "zip"
  source_content = "foobarbaz"
  source_content_filename = "file.txt"
  output_path = "${path.module}/files/zip.zip"
}

resource "aws_s3_bucket_object" "zip" {
  depends_on = ["data.archive_file.zip"]
  bucket = "steve-tf-test-bucket-20170719"
  key = "some-key"
  content = "${file(data.archive_file.zip.output_path)}"
  etag = "${md5(file(data.archive_file.zip.output_path))}"
}
```

### Debug Output
Second plan output: https://gist.github.com/ThePletch/383fe481e0e0b574952d323cea5c873f

### Expected Behavior
A second plan should list no diffs.

### Actual Behavior
First plan:
```
+ aws_s3_bucket_object.zip
    acl:                    "private"
    bucket:                 "steve-tf-test-bucket-20170719"
    content:                "PK\x03\x04\x14\x00\b\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\b\x00\x00\x00file.txtJ\xcb\xcfOJ,JJ\xac\x02\x04\x00\x00\xff\xffPK\a\b\xaa'x\x1a\x0f\x00\x00\x00\t\x00\x00\x00PK\x01\x02\x14\x00\x14\x00\b\x00\b\x00\x00\x00\x00\x00\xaa'x\x1a\x0f\x00\x00\x00\t\x00\x00\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00file.txtPK\x05\x06\x00\x00\x00\x00\x01\x00\x01\x006\x00\x00\x00E\x00\x00\x00\x00\x00"
    content_type:           "<computed>"
    etag:                   "1bf2d4bda2940b4c6271c385859ac756"
    key:                    "some-key"
    server_side_encryption: "<computed>"
    storage_class:          "<computed>"
    version_id:             "<computed>"

Plan: 1 to add, 0 to change, 0 to destroy.
```
Second plan:
```
~ aws_s3_bucket_object.zip
    content: "PK\x03\x04\x14\x00\b\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\b\x00\x00\x00file.txtJ��OJ,JJ�\x02\x04\x00\x00��PK\a\b�'x\x1a\x0f\x00\x00\x00\t\x00\x00\x00PK\x01\x02\x14\x00\x14\x00\b\x00\b\x00\x00\x00\x00\x00�'x\x1a\x0f\x00\x00\x00\t\x00\x00\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00file.txtPK\x05\x06\x00\x00\x00\x00\x01\x00\x01\x006\x00\x00\x00E\x00\x00\x00\x00\x00" => "PK\x03\x04\x14\x00\b\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\b\x00\x00\x00file.txtJ\xcb\xcfOJ,JJ\xac\x02\x04\x00\x00\xff\xffPK\a\b\xaa'x\x1a\x0f\x00\x00\x00\t\x00\x00\x00PK\x01\x02\x14\x00\x14\x00\b\x00\b\x00\x00\x00\x00\x00\xaa'x\x1a\x0f\x00\x00\x00\t\x00\x00\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00file.txtPK\x05\x06\x00\x00\x00\x00\x01\x00\x01\x006\x00\x00\x00E\x00\x00\x00\x00\x00"


Plan: 0 to add, 1 to change, 0 to destroy.
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Place the above config file on your machine (changing the bucket to an account owned bucket)
2. `terraform plan`
3. `terraform apply`
4. `terraform plan`

### Additional Notes
Probably important to note, since zip encoding differs between operating systems: this issue was present on both OSX Sierra and Ubuntu 14.04 (Trusty).

**Comment by jbardin** [comment]
Hi @ThePletch,

This is unfortunately due to how terraform stores its state data right now. Attributes and values are assumed to be utf8 strings, and unencoded binary data like the zip file will get altered during the serialization process.

There's probably not a satisfactory workaround, but you could ignore changes on the content and etag, and update the resource manually. You could alternatively base64 encode the content, but of course if that relies on other systems interfacing with the object they need to be able to un-encode it.

**Comment by quixand** [comment]
This is a bit of an annoying issue if you have lots of binary files being applied to s3. managing files like this in infrastructure seems like a basic requirement, I'm surprised this doesn't have higher priority. its confusing for teams as people incorrectly think there are changes to files.

Doesn't this bug effectively make the etag variable meaningless?

**Comment by ThePletch** [comment]
It seems like this can be worked around using the new `content_base64` attribute.

**Comment by quixand** [comment]
@ThePletch can you provide an implementation example please

**Comment by apparentlymart** [comment]
Indeed, the `content_base64` argument is part of the intended solution to this issue, though I think perhaps not all of the parts are in place yet for this to be fully solved in the general case, since we don't yet have a function to read a binary file from disk in base64 format. (That's planned for the next major release, along with [several other configuration language improvements](https://www.hashicorp.com/blog/terraform-0-1-2-preview).)

However, in the case of `aws_s3_bucket_object` it has an additional capability of accepting a file from local disk directly, in which case the provider itself reads the file (retaining the contents byte-for-byte) and the raw binary data doesn't need to pass through Terraform string variables (always unicode) at all:

```hcl
data "archive_file" "zip" {
  type                    = "zip"
  source_content          = "foobarbaz"
  source_content_filename = "file.txt"
  output_path             = "${path.module}/files/zip.zip"
}

resource "aws_s3_bucket_object" "zip" {
  bucket = "steve-tf-test-bucket-20170719"
  key    = "some-key"
  source = "${data.archive_file.zip.output_path}"
  etag   = "${data.archive_file.zip.output_md5}"
}
```

In the current release of Terraform. `path.module` expands to an absolute path and so unfortunately a configuration like the above isn't portable between different machines where the module is installed at different absolute paths; the provider will plan to update the object even though the content hasn't changed, because the plan no longer matches. We intend to address this in the next major release by making `path.module` return a path relative to the current working directory instead.

----

For the general problem here, moving forward we intend to standardize on base64 as the standard way to represent binary data within Terraform expressions, and so we'll be adding resource type arguments and functions as necessary to support reasonable binary-file use-cases. While in general we don't recommend passing around large binary objects in Terraform config (they need to be loaded into memory then), it's certainly reasonable to deal with small binary objects such as gzipping the result of rendering a template, and so we'll be gradually adding features as use-cases are discovered.

**Comment by apparentlymart** [comment]
Hi again, all!

With v0.12.0-alpha1 out I can make a more definitive followup to my earlier comment. We've added a new function [`filebase64`](https://github.com/hashicorp/terraform/blob/009534d1016cd3b1d0d911f151520b04f8db453a/website/docs/configuration/functions/filebase64.html.md) which can be used to read a binary file from disk into a base64-encoded string, which avoids the risk of corruption through misinterpretation as UTF-8. Along with this, the existing `file` function will now detect if it's asked to load something that _isn't_ UTF-8 and produce an explicit error recommending to use the other function, to remove the silent data corruption risk that was present before.

For `aws_s3_bucket_object` in particular I would recommend sticking with its `source` argument that reads directly from disk where possible, since this can also avoid the need to load the entire file content into memory and send it over the RPC channel to the provider plugin. However, for other cases where such a facility isn't available, such as "user data" on an EC2 instance, the new function fills the remaining functionality gap allowing binary data to be passed:

```hcl
resource "aws_instance" "example" {
  # ...

  user_data_base64 = filebase64("${path.module}/userdata.gz")
}
```

This combined with earlier changes to providers to accept data in base64 format should cover all the cases we know about. If you find a case where a resource type is trying to accept or produce binary data without using base64, please file an issue with that provider; base64 strings will now be the primary way to represent small binary objects in Terraform, while larger data should (as with `aws_s3_bucket_object`) be read directly from disk by the provider.

Since the new function is in master already, I'm going go close this out. The function is already available for experimental use in the alpha and will be included in the final v0.12.0 release in future. Thanks for reporting this, and sorry for the delay while we got all the separate parts in place for this.
